# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [47]:
import pandas as pd
df = pd.read_parquet('emendas_parlamentares (1).parquet')
df.head()

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.0,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.0,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.0,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.0,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.0,1131133.94,Sem informação


## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [48]:
df_ano_2022 = df[df['Ano da Emenda'] == 2022]
df_ano_2022.head(30)
df_ano_2022.shape

(6108, 29)

6108 emendas foram registradas em 2022.

## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [49]:
df_norte_500 = df[(df['Região'] == 'Norte') & (df['Valor Empenhado'] > 500000)]['Região'].value_counts()
print(df_norte_500.head())
df_norte_500.shape

Região
Norte    6230
Name: count, dtype: int64


(1,)

6752 emendas atendem as condições Norte e Valor Empenhado >= 500.000 ao mesmo tempo

## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [50]:
sp_rj = df[(df['UF'] == 'SÃO PAULO') | (df['UF'] == 'RIO DE JANEIRO')]['UF'].value_counts()
print(sp_rj)

UF
SÃO PAULO         10435
RIO DE JANEIRO     6988
Name: count, dtype: int64


O UF de São Paulo possui mais emendas, contendo 10.435. Enquanto isso, o UF do Rio de Janeiro possui apenas 6988 emendas em comparação.

## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [51]:
print('Antes:',df['Código da Emenda'])
df['Código da Emenda'] = df['Código da Emenda'].replace('Sem informação','Não Disponível')
print('Depois:',df['Código da Emenda'])

Antes: 0        Sem informação
1        Sem informação
2        Sem informação
3        Sem informação
4        Sem informação
              ...      
88654      201811420017
88655      201827690001
88656      201827690002
88657      201827690003
88658      201827690004
Name: Código da Emenda, Length: 88659, dtype: object
Depois: 0        Não Disponível
1        Não Disponível
2        Não Disponível
3        Não Disponível
4        Não Disponível
              ...      
88654      201811420017
88655      201827690001
88656      201827690002
88657      201827690003
88658      201827690004
Name: Código da Emenda, Length: 88659, dtype: object


## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [52]:
print('Antes:',df['Nome Função'].unique())
df_funcao = df['Nome Função'].str.upper()
print('Depois:',df_funcao.unique())

Antes: ['Urbanismo' 'Comércio e serviços' 'Saúde' 'Educação' 'Agricultura'
 'Assistência social' 'Desporto e lazer' 'Defesa nacional' 'Saneamento'
 'Direitos da cidadania' 'Gestão ambiental' 'Cultura'
 'Ciência e Tecnologia' 'Segurança pública' 'Organização agrária'
 'Comunicações' 'Administração' 'Habitação' 'Indústria' 'Trabalho'
 'Múltiplo' 'Transporte' 'Previdência social' 'Encargos especiais'
 'Relações exteriores' 'Energia' 'Essencial à justiça']
Depois: ['URBANISMO' 'COMÉRCIO E SERVIÇOS' 'SAÚDE' 'EDUCAÇÃO' 'AGRICULTURA'
 'ASSISTÊNCIA SOCIAL' 'DESPORTO E LAZER' 'DEFESA NACIONAL' 'SANEAMENTO'
 'DIREITOS DA CIDADANIA' 'GESTÃO AMBIENTAL' 'CULTURA'
 'CIÊNCIA E TECNOLOGIA' 'SEGURANÇA PÚBLICA' 'ORGANIZAÇÃO AGRÁRIA'
 'COMUNICAÇÕES' 'ADMINISTRAÇÃO' 'HABITAÇÃO' 'INDÚSTRIA' 'TRABALHO'
 'MÚLTIPLO' 'TRANSPORTE' 'PREVIDÊNCIA SOCIAL' 'ENCARGOS ESPECIAIS'
 'RELAÇÕES EXTERIORES' 'ENERGIA' 'ESSENCIAL À JUSTIÇA']


## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [53]:
duplicadas = df.duplicated().sum()
print(duplicadas)
df = df.drop_duplicates()
df.shape

63


(88596, 29)

Existem 63 linhas duplicadas no dataframe. Ao remover as duplicatas sobraram 88596 linhas.

## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro?

In [54]:
df_novo = df.drop_duplicates(subset=['Código da Emenda', 'Município'])
df_novo.shape

(75744, 29)

Sobraram 75744 linhas quando removemos as duplicatas considerando apenas as colunas Código da Emenda e Município. O resultado é diferente.

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [55]:
df.isnull().sum()

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,15961
Nome do Autor da Emenda,15961
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


Apenas as colunas Código do Autor da Emenda e Nome do Autor da Emenda possuem valores nulos, cada uma com 15961 valores nulos.

## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [56]:
print('Antes:',df['Código do Autor da Emenda'],df['Nome do Autor da Emenda'])
df['Código do Autor da Emenda'] = df['Código do Autor da Emenda'].fillna(-1)
df['Nome do Autor da Emenda'] = df['Nome do Autor da Emenda'].fillna('Desconhecido')
print('Depois:',df['Código do Autor da Emenda'],df['Nome do Autor da Emenda'])
df.isnull().sum()

Antes: 0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
          ...  
88591    4484.0
88592    4484.0
88593    4484.0
88594    4484.0
88595    4484.0
Name: Código do Autor da Emenda, Length: 88596, dtype: float64 0         None
1         None
2         None
3         None
4         None
         ...  
88591    ZUCCO
88592    ZUCCO
88593    ZUCCO
88594    ZUCCO
88595    ZUCCO
Name: Nome do Autor da Emenda, Length: 88596, dtype: object
Depois: 0          -1.0
1          -1.0
2          -1.0
3          -1.0
4          -1.0
          ...  
88591    4484.0
88592    4484.0
88593    4484.0
88594    4484.0
88595    4484.0
Name: Código do Autor da Emenda, Length: 88596, dtype: float64 0        Desconhecido
1        Desconhecido
2        Desconhecido
3        Desconhecido
4        Desconhecido
             ...     
88591           ZUCCO
88592           ZUCCO
88593           ZUCCO
88594           ZUCCO
88595           ZUCCO
Name: Nome do Autor da Emenda, Length: 8

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,0
Nome do Autor da Emenda,0
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [57]:
print(df['Ano da Emenda'].dtype)
df['Ano da Emenda'] = df['Ano da Emenda'].astype(str)
print(df['Ano da Emenda'].dtype)

int64
object


O tipo da coluna Ano da Emenda era int64. Após a conversão, vira do tipo string.

## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [58]:
print(df['Valor Empenhado'].dtype)
df['Valor Emepnhado'] = df['Valor Empenhado'].astype(float)
print(df['Valor Empenhado'].dtype)

float64
float64


A coluna já era do tipo float64, então após a conversão, ainda fica como float64.

## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [59]:
valor_total_regiao = df.groupby('Região')['Valor Pago'].sum().sort_values(ascending=False)
print(valor_total_regiao.head(2))

Região
Nordeste    4.644279e+10
Sudeste     3.426218e+10
Name: Valor Pago, dtype: float64


As duas regiões que mais receberam recursos são a região Nordeste e a região Sudeste.

## 13) Calcule a **quantidade** de emendas por **Nome Função**. Mostre as 5 funções com mais emendas.

In [60]:
df['Nome Função'].value_counts().head()

,count
Nome Função,
Saúde,37039
Urbanismo,8655
Educação,8101
Agricultura,4938
Encargos especiais,4610


## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [61]:
df['Saldo a Pagar'] = df['Valor Empenhado'] - df['Valor Pago']
print(df[['Valor Empenhado', 'Valor Pago', 'Saldo a Pagar']].head())

   Valor Empenhado  Valor Pago  Saldo a Pagar
0        700000.00         0.0      700000.00
1        487500.00         0.0      487500.00
2        100000.00         0.0      100000.00
3        243750.00         0.0      243750.00
4       1658333.94         0.0     1658333.94


## 15) Elimine a(s) coluna(s) do dataframe que eestá totalmente vazia.

In [62]:
df = df.drop(columns=['Código da Emenda_1'])


## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.

In [65]:
df.to_json('emendas_transformadas.json', index=False)